pkg/runtime/optimize/hints.go

```diff
@@ -0,0 +1,113 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package optimize
+
+import (
+	"github.com/pkg/errors"
+)
+
+import (
+	"github.com/arana-db/arana/pkg/proto/hint"
+	"github.com/arana-db/arana/pkg/proto/rule"
+	"github.com/arana-db/arana/pkg/runtime/ast"
+)
+
+func init() {
+	RegisterHint(hint.TypeFullScan, &FullScan{})
+	RegisterHint(hint.TypeDirect, &Direct{})
+}
+
+type HintExecutor interface {
+	shard(tableName ast.TableName, rule *rule.Rule) (shards rule.DatabaseTables, err error)
+}
+
+var (
+	hintHandlers = make(map[hint.Type]HintExecutor)
+)
+
+// RegisterHint register hint implementation.
+func RegisterHint(t hint.Type, h HintExecutor) {
+	hintHandlers[t] = h
+}
+
+// validate hints of the same type are allowed only once,such as (master||slave) || (router||fullScan||direct)
+func validate(hints []*hint.Hint) error {
+	shardingType := make(map[bool]hint.Type, 1)
+	nodeType := make(map[bool]hint.Type, 1)
+	for _, v := range hints {
+		if v.Type == hint.TypeFullScan || v.Type == hint.TypeDirect || v.Type == hint.TypeRoute {
+			if exist, ok := shardingType[true]; ok {
+				return errors.Errorf("hint type conflict:%s,%s", exist.String(), v.Type.String())
+			}
+			shardingType[true] = v.Type
+		}
+		if v.Type == hint.TypeMaster || v.Type == hint.TypeSlave {
+			if exist, ok := nodeType[true]; ok {
+				return errors.Errorf("hint type conflict:%s,%s", exist.String(), v.Type.String())
+			}
+			nodeType[true] = v.Type
+		}
+	}
+	return nil
+}
+
+func Hints(tableName ast.TableName, hints []*hint.Hint, rule *rule.Rule) (shards rule.DatabaseTables, err error) {
+	err = validate(hints)
+	if err != nil {
+		return nil, err
+	}
+	var (
+		executor HintExecutor
+		ok       bool
+	)
+	for _, v := range hints {
+		if executor, ok = hintHandlers[v.Type]; ok {
+			break
+		}
+	}
+	if executor == nil {
+		return nil, nil
+	}
+	shards, err = executor.shard(tableName, rule)
+	if err != nil {
+		return nil, err
+	}
+	return shards, nil
+}
+
+// FullScan hint fullScan
+type FullScan struct {
+}
+
+func (h *FullScan) shard(tableName ast.TableName, rule *rule.Rule) (shards rule.DatabaseTables, err error) {
+	return rule.MustVTable(tableName.Suffix()).Topology().Enumerate(), nil
+}
```

Comment:
Incorrect logic, full-scan affects only if the shard results is nil. In other words, the full-scan HINT works after shard computing.

Reply:
done